### PR TITLE
move HTTP-specific options from `TRPCClient` to links

### LINKED ISSUE #2709: Move HTTP-specific things from runtime to links
tRPC isn't tied to HTTP so we shouldn't actually enforce using HTTP when creating the TRPC client.

This could be done by deprecating those fields like the below, but then we have no _guarantee_ when creating the HTTPLinks that you do it right.

```diff
interface CreateTRPCClientBaseOptions {
  /**
   * Add ponyfill for fetch
+   * @deprecated moved to HTTP Links
   */
  fetch?: typeof fetch;
  /**
+   * @deprecated moved to HTTP Links
   * Add ponyfill for AbortController
   */
  AbortController?: typeof AbortController;
  /**
   * Headers to be set on outgoing requests or a callback that of said headers
   * @link http://localhost:3000/docs/v10/header
+   * @deprecated moved to HTTP Links
   */
  headers?: HTTPHeaders | (() => HTTPHeaders | Promise<HTTPHeaders>);
  /**
   * Data transformer
   * @link https://trpc.io/docs/data-transformers
   **/
  transformer?: ClientDataTransformerOptions;
}
```

And

```diff

export interface HTTPLinkOptions {
  url: string;
+  fetch?: TRPCFetch;
+  AbortController?: typeof AbortController;
+  headers?: () => HTTPHeaders | Promise<HTTPHeaders>;
}
```

_Originally posted by @KATT in https://github.com/trpc/trpc/issues/2700#issuecomment-1245930214_

### PR DESCRIPTION
**Pull request by KATT**
Closes #2709, cc @moltar

## 🎯 Changes

- Removes HTTP-specific options from the `TRPCClient` and moves them to links
- Enables more transport protocols w/o hacking around this
- Makes bundle size smaller for ppl not using `httpBatchLink` as it's not always imported
- Breaking change

**Docs:** https://www-git-issues-2709-trpc.vercel.app/docs/v10/migrate-from-v9-to-v10#http-specific-options-moved-from-trpcclient-to-links

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [ ] If necessary, I have added documentation related to the changes made.
- [ ] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/client/src/createTRPCClient.ts [hunk 1]
```diff
@@ -5,22 +5,52 @@ import type { AnyRouter } from '@trpc/server';
...
+        headers?: HTTPLinkOptions['headers'];
+        /**
+         * @deprecated use `links` instead
+         */
+        fetch?: HTTPLinkOptions['fetch'];
+        /**
+         * @deprecated use `links` instead
+         */
+        AbortController?: HTTPLinkOptions['AbortController'];
+      },
```
Comment (KATT) [inline packages/client/src/createTRPCClient.ts:36]: Makes sort-of backwards compat

**Review by juliusmarminge — APPROVED** [review]

File: packages/client/src/createTRPCClient.ts [hunk 2]
```diff
@@ -1,26 +1,57 @@
...
+         */
+        headers?: HTTPLinkOptions['headers'];
+        /**
+         * @deprecated use `links` instead
+         */
+        fetch?: HTTPLinkOptions['fetch'];
+        /**
+         * @deprecated use `links` instead
+         */
+        AbortController?: HTTPLinkOptions['AbortController'];
```
[the remark this answers is not in the document]
Comment (KATT) [reply packages/client/src/createTRPCClient.ts:35]: Feel free to remove it and have a stab at refactoring if you want 😄

**Review by jlalmes — APPROVED** [review]

File: packages/client/src/createTRPCClientProxy.ts [hunk 3]
```diff
@@ -21,6 +21,7 @@ import { TRPCClientError } from './TRPCClientError';
 import { CreateTRPCClientOptions, createTRPCClient } from './createTRPCClient';
```
Comment (moltar) [inline packages/client/src/createTRPCClientProxy.ts:21]: Lint issue

File: packages/client/src/internals/TRPCClient.ts [hunk 4]
```diff
@@ -2,6 +2,7 @@ import {
   AnyRouter,
   ClientDataTransformerOptions,
   DataTransformer,
+  ProcedureRecord,
```
Comment (moltar) [inline packages/client/src/internals/TRPCClient.ts:5]: Lint

**Review by moltar — COMMENTED** [review]
Thanks! LGTM. Just a few linting issues, not sure if you care about those. But solution wise, I think it's a good direction!

One other thing I'd like to see is fetch-compatible interface being tRPC-driven, rather than `typeof fetch`. Maybe I can take a stub at this?

**Comment by juliusmarminge** [comment]
> Just a few linting issues, not sure if you care about those

I've fixed those as part of #2726

**Pull request merged**